### DISCUSSION TITLE
Bug in IF conditions on inputs for reusable workflows

### DISCUSSION THREAD

**Discussion by stevengonsalvez** [community/community#25293] [Actions]
<h2>
<a name="problem-1" class="anchor" href="#problem-1"></a>Problem:</h2>
<p>When a reusable workflow (called workflow), contains if conditions (based on the input of the job) at job level, there is a observation that the if condition seems to execute and use value globally (causing jobs which are being skipped because of a check).</p>
<h3>
<a name="example-2" class="anchor" href="#example-2"></a>Example</h3>
<blockquote>
<p>Caller workflow: Simple workflow calling a reusable workflow passing in the github event (PR or push) to the reusable workflow</p>
</blockquote>
<pre><code class="lang-yaml">name: caller-workflow
on:
  pull_request:
  push:
    branches:
      - master
# 2
jobs:
  call-the-workflow:
    uses: action-foobar/action-testing/.github/workflows/called_workflow.yml@master
    with:
      TRIGGER_EVENT : ${{ github.event_name }}
</code></pre>
<blockquote>
<p>Reusable workflow: Workflow with a sequence of job which has conditions based on the input (whether it is a pull_request or a push)</p>
</blockquote>
<pre><code class="lang-yaml">name: tf-app-dns-reusable

on:
  workflow_call:
    inputs:
      TRIGGER_EVENT:
        description: trigger event for the workflow
        required: true
        type: string

jobs:
  lint:
    runs-on: ubuntu-latest
    name: Validate terraform configuration

    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: debug
        run: |
          echo ${{ inputs.TRIGGER_EVENT }}
  plan-non-prod:
    if: ${{ inputs.TRIGGER_EVENT == 'pull_request' }}
    needs: lint
    runs-on: ubuntu-latest
    name: Plan Application DNS Non Prod
    steps:
      - name: debug
        run: |
          echo "triggered event"
          echo ${{ inputs.TRIGGER_EVENT }}
  plan-prod:
    if: ${{ inputs.TRIGGER_EVENT == 'pull_request' }}
    needs: "plan-non-prod"
    runs-on: ubuntu-latest
    name: Plan Application DNS Prod
    steps:
      - name: debug
        run: |
          echo "triggered event"
          echo ${{ inputs.TRIGGER_EVENT }}
  apply-non-prod:
    if: ${{ inputs.TRIGGER_EVENT == 'push' }}
    needs: lint
    runs-on: ubuntu-latest
    name: apply non prod
    steps:
      - name: debug
        run: |
          echo "triggered event"
          echo ${{ inputs.TRIGGER_EVENT }}
</code></pre>
<h4>
<a name="observation-3" class="anchor" href="#observation-3"></a>Observation</h4>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/2/e/2e6721fb54975db544836c805c7f056d66abf1bd.png" data-download-href="/uploads/short-url/6CuVjQW19TpkILaFtBf7UZGkT8h.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/9320602/181090107-256c4816-71a4-47a0-bd20-8bebf3bca6b3.png" alt="image" data-base62-sha1="6CuVjQW19TpkILaFtBf7UZGkT8h" width="690" height="195" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/2/e/2e6721fb54975db544836c805c7f056d66abf1bd_2_690x195.png, https://aws1.discourse-cdn.com/github/optimized/3X/2/e/2e6721fb54975db544836c805c7f056d66abf1bd_2_1035x292.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/2/e/2e6721fb54975db544836c805c7f056d66abf1bd.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/2/e/2e6721fb54975db544836c805c7f056d66abf1bd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1182×335 58.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<ul>
<li>
<p>Job 3 gets the check skipped , even though the condition is valid and matches as Job 2 (Job 3 only depends on success of job 2 and the same <code>IF</code> condition as job2.</p>
</li>
<li>
<p>On removal of the <code>IF</code> condition on Job 4 , then Job 3 executes fine .</p>
</li>
</ul>
<p>Which implies of some sort of “state” of the IF condition (since there is no relationship between Job 4 and Job 3).</p>
<blockquote>
<p>PS: This was working perfectly fine till  11PM GMT 10th Jan . (as we have a lot of jobs on this setup)</p>
</blockquote>
<p>The observation of runs can be found in this repository here → <a href="https://github.com/action-foobar/action-testing/actions/runs/1682109402" class="inline-onebox" rel="noopener nofollow ugc">push stuff · action-foobar/action-testing@500bca2 · GitHub</a></p>

**Comment by AlexFernandes-MOVAI** [comment]
<p>Same issue seen on our pipelines here:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/MOV-AI/.github/blob/a20d4764cadcdaa8f206dc247255d5ee28dea354/.github/workflows/reusable-ros-build-pack.yml#L141">
  <header class="source">

      <a href="https://github.com/MOV-AI/.github/blob/a20d4764cadcdaa8f206dc247255d5ee28dea354/.github/workflows/reusable-ros-build-pack.yml#L141" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/MOV-AI/.github/blob/a20d4764cadcdaa8f206dc247255d5ee28dea354/.github/workflows/reusable-ros-build-pack.yml#L141" target="_blank" rel="noopener nofollow ugc">MOV-AI/.github/blob/a20d4764cadcdaa8f206dc247255d5ee28dea354/.github/workflows/reusable-ros-build-pack.yml#L141</a></h4>



  <pre class="onebox">    <code class="lang-yml">
      <ol class="start lines" start="131" style="counter-reset: li-counter 130 ;">
          <li>
          </li>
<li>- name: Print generated packages</li>
          <li>  run: |</li>
          <li>    ls -la artifacts</li>
          <li>
          </li>
<li>- name: Print generated rosdep</li>
          <li>  run: |</li>
          <li>    cat /usr/local/rosdep/ros-pkgs.yaml</li>
          <li>
          </li>
<li>- name: Rosdep Publish</li>
          <li class="selected">  if: ${{ inputs.deploy == 'true' }}</li>
          <li>  run: |</li>
          <li>    export AWS_DEFAULT_REGION="eu-west-1"</li>
          <li>    export AWS_ACCESS_KEY_ID="${{ secrets.aws_sqs_rosdep_access_key }}"</li>
          <li>    export AWS_SECRET_ACCESS_KEY='${{ secrets.aws_sqs_rosdep_secret_access_key }}'</li>
          <li>    mobros publish</li>
          <li>    cat /usr/local/rosdep/ros-pkgs.yaml</li>
      </ol>
    </code>
  </pre>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by Kerizer** [comment]
<p>Faced same issue, our CI/CD is pretty much blocked now</p>

**Comment by duartecoelhomovai** [comment]
<p>I’ve even removed the IF, and the second job continues to be skipped. We are really blocked!</p>
<p><img src="https://user-images.githubusercontent.com/84720623/181090109-74293d12-4321-4ece-939e-23aa9706a321.png" alt="Screenshot from 2022-01-11 14-50-44" data-base62-sha1="dom5ANp7v2d5qTAEbGQ4igrAynu" width="632" height="234"></p>

**Comment by wongjn** [comment]
<p>Seems to happen when a sibling job gets skipped, any subsequent jobs then get skipped, even if they are not related.</p>
<p>See run <a href="https://github.com/wongjn/test/actions/runs/1682106999" class="inline-onebox" rel="noopener nofollow ugc">Create main.yml · wongjn/test@f05463b · GitHub</a>.<br>
See that <em>Testing/test</em> job is skipped even though the dependent job, <em>Testing/start</em> was successful.<br>
This happens because <em>Testing/assets</em> has been skipped via the input <code>assets</code> being <code>false</code> . Setting <code>assets</code> to <code>true</code> will now mean <a href="https://github.com/wongjn/test/actions/runs/1682140972" rel="noopener nofollow ugc"> <em>Testing/test</em> now runs</a> because <em>Testing/assets</em> is <strong>not</strong> skipped. This seems to be an erroneous coupling of these jobs even though the graph indicates there should be no dependency.</p>

**Comment by duartecoelhomovai** [comment]
<p>Thanks for the info. With this we are able to workaround temporarily. Also opened a github ticket pointing to this thread to see if we raise the criticality on this.</p>

**Comment by stevengonsalvez** [comment]
<p>Watch this issue - <a href="https://github.com/actions/runner/issues/1602" class="inline-onebox" rel="noopener nofollow ugc">Bug in IF conditions on inputs for reusable workflows · Issue #1602 · actions/runner · GitHub</a></p>

**Comment by jacek-jablonski** [comment]
<p><a class="mention" href="/surveyforcustomers">@Surveyforcustomers</a> invalid link</p>

**Comment by gustavocoleta** [comment]
<p>If i realy understend, you need change the job 3 like this:</p>
<pre><code class="lang-auto">plan-prod:
    if: ${{ always() }} &amp;&amp; ${{needs.plan-non-prod.outputs.status == 'success'    
    needs: "plan-non-prod"
    runs-on: ubuntu-latest
    name: Plan Application DNS Prod
...
</code></pre>
<p>Will run <code>always</code> and check output status of <code>plan-non-prod</code> is <code>success</code>. But i don’t test is, i have some like this</p>

**Comment by airtower-luna** [comment]
<p>You need to wrap the whole expression in <code>${{ ... }}</code> or leave them off entirely:</p>
<pre><code class="lang-yaml">    if: ${{ always() &amp;&amp; needs.plan-non-prod.outputs.status == 'success' }}
</code></pre>
<p>With partial expressions the output is just a string (e.g. “true &amp;&amp; false”) and a non-empty string evaluates to <code>true</code>.</p>

**Comment by stevengonsalvez** [comment, marked as answer]
<p>This is resolved from <a href="https://github.com/actions/runner/issues/1602" class="inline-onebox" rel="noopener nofollow ugc">Bug in IF conditions on inputs for reusable workflows · Issue #1602 · actions/runner · GitHub</a> .</p>
